File: src/leetcode/stack/key-value-store-nested-transactions.ts
// DIFFICULTY: MEDIUM
//
// Implement a key-value store with the following functionality:
//
// - get(key: string)
// - set(key: string, value: number)
// - delete(key: string)
// - begin() starts a transaction
// - commit() commits a transaction
// - rollback() rolls back a transaction
//
// This structure supports nested transactions, and the topmost transaction should be able to see the values in the
// transactions below it.
//
// NOTE: This is not an official LeetCode question, but it was asked by Cruise and Bloomberg (sometimes with nested
// transactions and sometimes without).
//
// See {@link https://leetcode.com/discuss/interview-question/279913/Bloomberg-or-Onsite-or-Key-Value-Store-with-transactions}
export { KeyValueStore };

// SOLUTION:
//
// The key to this problem is to maintain a stack of maps.  When you start a transaction, you push a new map onto the
// stack.
//
// To improve runtime performance, we maintain a current transaction context that gets propagated to the top of the
// stack (or thrown away if we rollback).
class KeyValueStore {
  private readonly stack: Map<string, number | null>[];

  private readonly map: Map<string, number>;

  constructor() {
    this.stack = [];
    this.map = new Map();
  }

  public get(key: string): number | undefined {
    if (this.stack.length === 0) {
      return this.map.get(key);
    }

    let frame = this.stack.length - 1;
    while (frame >= 0) {
      if (this.stack[frame].has(key)) {
        return this.stack[frame].get(key) ?? undefined;
      }

      frame--;
    }

    return undefined;
  }

  public set(key: string, value: number): void {
    if (this.stack.length === 0) {
      this.map.set(key, value);
      return;
    }

    this.stack[this.stack.length - 1].set(key, value);
  }

  public delete(key: string): void {
    if (this.stack.length === 0) {
      this.map.delete(key);
      return;
    }

    this.stack[this.stack.length - 1].set(key, null);
  }

  public begin(): void {
    this.stack.push(new Map());
  }

  public commit() {
    if (this.stack.length === 0) {
      return;
    }

    const target = this.stack.length === 1 ? this.map : this.stack[this.stack.length - 2];
    const source = this.stack.pop()!;

    for (const [key, value] of source) {
      if (value === null) {
        target.delete(key);
        continue;
      }

      target.set(key, value);
    }
  }

  public rollback() {
    if (this.stack.length === 0) {
      return;
    }

    const map = this.stack.pop()!;
    map.clear();
  }
}

File: src/leetcode/stack/key_value_store_nested_transactions.py
# DIFFICULTY: MEDIUM
# ------------------
#
# Implement a key-value store with the following functionality:
#
# - get(key: string)
# - set(key: string, value: number)
# - delete(key: string)
# - begin() starts a transaction
# - commit() commits a transaction
# - rollback() rolls back a transaction
#
# This structure supports nested transactions, and the topmost transaction should be able to see the values in the
# transactions below it.
#
# See https://leetcode.com/discuss/interview-question/279913/Bloomberg-or-Onsite-or-Key-Value-Store-with-transactions
class KeyValueStore:
    """
    SOLUTION
    --------

    Use a stack to store the transactions.  Each transaction is a dictionary that maps keys to values.

    COMPLEXITY
    ----------

    Time complexity is O(1) for all operations.

    Space complexity is O(n) where n is the number of keys in the data structure.
    """

    def __init__(self) -> None:
        self.deletions: list[set[str]] = []
        self.txs: list[dict[str, int]] = []
        self.main: dict[str, int] = {}

    def get(self, key: str) -> int | None:
        if not self.txs:
            return self.main.get(key, None)

        frame = len(self.txs) - 1
        while frame >= 0:
            if key in self.deletions[frame]:
                return None
            if key in self.txs[frame]:
                return self.txs[frame].get(key, None)
            frame -= 1

        return None

    def set(self, key: str, value: int) -> None:
        if not self.txs:
            self.main[key] = value
            return

        self.txs[-1][key] = value

    def delete(self, key: str) -> None:
        if not self.txs:
            self.main.pop(key, None)
            return

        self.txs[-1].pop(key, None)
        self.deletions[-1].add(key)

    def begin(self) -> None:
        self.txs.append({})
        self.deletions.append(set())

    def commit(self) -> None:
        if not self.txs:
            return

        target = self.main if len(self.txs) == 1 else self.txs[-2]
        source = self.txs[-1]

        for key, value in source.items():
            target[key] = value

        for key in self.deletions[-1]:
            target.pop(key, None)

        self.txs.pop()
        self.deletions.pop()

    def rollback(self) -> None:
        if not self.txs:
            return

        self.txs.pop()
        self.deletions.pop()
